drivers/flash/spi_flash.c

```diff
@@ -0,0 +1,523 @@
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
+	hdr[i++] = offset >> 8;
+	hdr[i++] = offset;
+	r = spi_transceive(&data->config, txbufs, ARRAY_SIZE(txbufs), NULL, 0);
+	if (r == 0) {
+		wait_for_flash_idle(dev);
+	}
+	k_sem_give(&data->sem);
+
+err:
+	return r;
+}
+
+static int spi_flash_write_protection_set(struct device *dev, bool enable)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	int r;
+
+	k_sem_take(&data->sem, K_FOREVER);
+	wait_for_flash_idle(dev);
+	if (enable) {
+		r = spi_flash_cmd(dev, CMD_WRITE_DISABLE);
+	} else {
+		SYS_LOG_DBG("global unblock %02x", data->global_unprotect_opcode);
+
+		r = spi_flash_cmd(dev, CMD_WRITE_ENABLE);
+		if (r == 0 && data->global_unprotect_opcode) {
+			wait_for_flash_idle(dev);
+			r = spi_flash_cmd(dev, data->global_unprotect_opcode);
+		}
+	}
+	k_sem_give(&data->sem);
+
+	return r;
+}
+
+static inline int spi_flash_erase_cmd(struct device *dev, u8_t opcode, off_t offset)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	u8_t hdr[5];
+	struct spi_buf txbufs[] = {
+		{ .buf = hdr,  .len = 1 + data->addr_len },
+	};
+	int r, i;
+
+	SYS_LOG_DBG("erase: %02x %06x", opcode, offset);
+
+	wait_for_flash_idle(dev);
+	r = spi_flash_cmd(dev, CMD_WRITE_ENABLE);
+	if (r != 0) {
+		return r;
+	}
+
+	i = 0;
+	hdr[i++] = opcode;
+	if (data->addr_len == 4) {
+		hdr[i++] = offset >> 24;
+	}
+	hdr[i++] = offset >> 16;
+	hdr[i++] = offset >> 8;
+	hdr[i++] = offset;
+
+	wait_for_flash_idle(dev);
+	return spi_transceive(&data->config, txbufs, ARRAY_SIZE(txbufs), 0, 0);
+}
+
+static int spi_flash_erase(struct device *dev, off_t offset, size_t size)
+{
+	struct spi_flash_data *const data = dev->driver_data;
+	off_t offs = offset;
+	size_t left = size;
+	int ret = 0;
+
+	if (offset < 0 || offset+size >= data->flash_size ||
+	    (offset & 0xfff) || (size & 0xfff)) {
+		return -ENODEV;
+	}
+	if (!data->bfp.opcode_erase_4k) {
+		return -ENOTSUP;
+	}
+
+	k_sem_take(&data->sem, K_FOREVER);
+
+	/* FIXME: Support chip erase, and large block erase */
+
+	/* Erase in 4K blocks */
+	while (left && ret == 0) {
+		ret = spi_flash_erase_cmd(dev, data->bfp.opcode_erase_4k, offs);
+		offs += 0x1000;
+		left -= 0x1000;
+	}
+
+	k_sem_give(&data->sem);
+	return ret;
+}
+
+static const struct flash_driver_api spi_flash_api = {
+	.read = spi_flash_read,
+	.write = spi_flash_write,
+	.erase = spi_flash_erase,
+	.write_protection = spi_flash_write_protection_set,
+#if defined(CONFIG_FLASH_PAGE_LAYOUT)
+	.page_layout = (flash_api_pages_layout)
+		       flash_page_layout_not_implemented,
+#endif
+	.write_block_size = 1,
+};
+
+static int spi_flash_init(struct device *dev)
+{
+	struct device *spi_dev;
+	struct spi_flash_data *data = dev->driver_data;
+	int r;
+
+	spi_dev = device_get_binding(CONFIG_SPI_FLASH_SPI_NAME);
+	if (!spi_dev) {
+		return -EIO;
+	}
+
+	data->config = (struct spi_config) {
+		.dev = spi_dev,
+		.frequency = CONFIG_SPI_FLASH_SPI_FREQ_0,
+		.operation = SPI_WORD_SET(8),
+	};
+
```

Comment:
You will need to add support for gpio based CS. See what I did, for instance, in https://github.com/zephyrproject-rtos/zephyr/pull/5217/commits/3caa3ba293dbe5954c84a394d906865cf9d609c6 line 746.